### PR TITLE
feat: Merge copywriter with commands

### PR DESCRIPTION
**Pull request by gsppe**
## Description

Unifies the operations AI endpoint with the copywriter feature. For copy we do a LLM call to detect if the prompt is a copywriter request and if so we use the copywriter chain.

- [x] Depends on #2391

## Steps for reproduction

1. Using the commands tool ask to write, rewrite or translate some copy. 
2. Try to give other non-text generation commands to make sure the detection call works fine

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - test it on preview
- [ ] @TrySound  detailed review (read every line)

## Before requesting a review

- [x] made a self-review

## Before merging

- [x] tested locally and on preview environment (preview dev login: 5de6)

### CODE REVIEW

File: apps/builder/app/builder/features/ai/operations.tsx [hunk 1]
```diff
@@ -58,7 +65,12 @@ export const Operations = () => {
...
         }
 
-        if (typeof prompt !== "string" || typeof data !== "string") {
+        if (
+          typeof prompt !== "string" ||
+          typeof data !== "string" ||
+          typeof projectId !== "string" ||
+          typeof instanceId !== "string"
+        ) {
           alert("Invalid data");
```
Comment (kof) [inline apps/builder/app/builder/features/ai/operations.tsx:74]: do you want to merge alert?

File: apps/builder/app/builder/features/ai/operations.tsx [hunk 2]
```diff
@@ -74,18 +86,43 @@ export const Operations = () => {
 
         abort.current = new AbortController();
 
-        request<operations.Response>([
-          restAiOperations(),
+        request<operations.Response>(
```
Comment (kof) [inline apps/builder/app/builder/features/ai/operations.tsx:89]: this whole logic really shouldn't be inside the component, but a function above

File: apps/builder/app/routes/rest.ai.op.ts [hunk 3]
```diff
@@ -5,17 +5,31 @@ import {
...
+    instanceId: z.string(),
+  })
+);
 
 export const maxDuration = 120;
 
 export const action = async ({ request }: ActionArgs) => {
+  const llmMessages: ModelMessage[] = [];
+
   if (isFeatureEnabled("aiCopy") === false) {
```
Comment (kof) [inline apps/builder/app/routes/rest.ai.op.ts:33]: I think we are launching most of these AI features at once, so probably no reason to separate into separate feature flags

File: packages/ai/src/chains/copywriter/chain-copywriter-detect.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    const llmMessages: ModelMessage[] = [
+      [
+        "assistant",
+        'Given a user request determine whether it is a text generation, translation or editing operation or not. Respond exclusively with "yes" or "no"',
+      ],
+      ["user", formatPrompt({ prompt }, "Request:\n```{prompt}```")],
+    ];
+
+    const messages = model.generateMessages(llmMessages);
+    const completion = await model.completion({
```
Comment (kof) [inline packages/ai/src/chains/copywriter/chain-copywriter-detect.ts:42]: tw how quick is this call on average in the end? just curious

File: apps/builder/app/builder/features/ai/operations.tsx (hunk 1 above)
Comment (kof) [reply apps/builder/app/builder/features/ai/operations.tsx:74]: maybe add @todo statements to be able to search for things to fix

File: apps/builder/app/builder/features/ai/apply-operations.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    metas,
+    baseBreakpoint.id
+  );
+
+  // Occasionally the LLM picks a component name as the insertion point.
+  // Instead of throwing the otherwise correct operation we try to fix this here.
+  if (
+    [...metas.keys()].some((ComponentName) =>
+      ComponentName.includes(operation.addTo)
+    )
```
Comment (TrySound) [inline apps/builder/app/builder/features/ai/apply-operations.ts:72]: Could you clarify why you check substring in component name?
And please start it with lower case

File: packages/ai/src/index.ts [hunk 6]
```diff
@@ -1,15 +1,8 @@
...
-export {
-  create as createGptModel,
-  type Model as GPTModel,
-  type ModelConfig as GPTModelConfig,
-  type ModelMessageFormat as GPTModelMessageFormat,
-} from "./models/gpt";
+export * as commandDetect from "./chains/command-detect/index.server";
+export * as copywriter from "./chains/copywriter/index.server";
+export * as operations from "./chains/operations/index.server";
+export * as templateGenerator from "./chains/template-generator/index.server";
```
Comment (TrySound) [inline packages/ai/src/index.ts:5]: Isn't the point of client entry point to not export server deps?

File: packages/ai/src/chains/operations/index.ts [hunk 7]
```diff
@@ -1,7 +1,62 @@
...
+  wsOperation as editStylesWsOperation,
+} from "./edit-styles";
+export const editStyles = {
+  aiOperation: editStylesAiOperation,
+  wsOperation: editStylesWsOperation,
+};
+export type editStyles = {
+  aiOperation: editStylesAiOperation;
+  wsOperation: editStylesWsOperation;
+};
```
[the remark this answers is not in the document]
Comment (TrySound) [reply packages/ai/src/chains/operations/index.ts:23]: idk, don't use this style? Ecosystem moving away from namespaces.

File: apps/builder/app/builder/features/ai/commands-bar.tsx [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+      engine.addStyleRule(`[${idAttribute}="${instanceId}"]${state ?? ""}`, {
+        breakpoint: breakpointId,
+        style,
+      });
+    }
+
+    return `<style>{\`${engine.cssText.replace(/\n/gm, " ")}\`}</style>${jsx
+      .replace(new RegExp(`${idAttribute}="[^"]+"`, "g"), "")
+      .replace(/\n(data-)/g, " $1")}`;
+  }
```
Comment (TrySound) [inline apps/builder/app/builder/features/ai/commands-bar.tsx:272]: How styles work without id?

File: packages/feature-flags/src/flags.ts [hunk 9]
```diff
@@ -6,5 +6,4 @@ export const displayContents = false;
 export const adminRole = false;
 // A general flag to enable/disable all the AI features.
 export const ai = process.env.NODE_ENV !== "production";
```
Comment (TrySound) [inline packages/feature-flags/src/flags.ts:8]: Please replace NODE_ENV check with false. In dev we show features by settings FEATURES=* in .env or in localstorage

**Review by TrySound — APPROVED** [review]

File: apps/builder/app/builder/features/ai/commands-bar.tsx [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+import { RequestParamsSchema } from "~/routes/rest.ai";
+import untruncateJson from "untruncate-json";
+import { traverseTemplate } from "@webstudio-is/jsx-utils";
+
+const handleSubmit = async (
+  event: React.FormEvent<HTMLFormElement>,
+  abortSignal: AbortSignal
+) => {
+  const requestParams = Object.fromEntries(
+    new FormData(event.currentTarget).entries()
```
Comment (kof) [inline apps/builder/app/builder/features/ai/commands-bar.tsx:40]: I think .entries() call happens automatically when formdata is passed to fromEntries

File: apps/builder/app/builder/features/ai/commands-bar.tsx [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+    }
+  );
+
+  if (result.success === true) {
+    if (result.type === "json" && result.id === "operations") {
+      restoreComponentsNamespace(result.data);
+      applyOperations(result.data);
+    }
+    return;
+  } else {
```
Comment (kof) [inline apps/builder/app/builder/features/ai/commands-bar.tsx:90]: this else is not needed after a return

File: apps/builder/app/builder/features/ai/commands-bar.tsx [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+          node.component =
+            "@webstudio-is/sdk-components-react-radix:" +
+            node.component.slice("Radix.".length);
+        }
+      });
+    }
+  }
+};
+
+const $jsx = computed(
```
Comment (kof) [inline apps/builder/app/builder/features/ai/commands-bar.tsx:197]: that is so powerful!

File: packages/feature-flags/src/flags.ts (hunk 9 above)
Comment (kof) [reply packages/feature-flags/src/flags.ts]: its enabled in preview, we have there asterix as a value that enables all features, recommend setting the same in your .env locally

**Pull request merged**